### PR TITLE
[SIR] use datatype of literal node (issue #612)

### PR DESCRIPTION
**Pull request by rupertford**

### CODE REVIEW

**Comment by rupertford** [comment]
Ready for review. Probably most suited to @arporter or @sergisiso.

File: src/psyclone/psyir/backend/sir.py [hunk 1]
```diff
@@ -437,10 +450,16 @@ def unaryoperation_node(self, node):
                 isinstance(node.children[0], Literal)):
             raise VisitorError(
                 "Currently, unary operators can only be applied to literals.")
-        result = node.children[0].value
-        # There is an assumption here that the literal is a float (see #612)
-        return ("{0}make_literal_access_expr(\"{1}{2}\", BuiltinType.Float)"
-                "".format(self._nindent, oper, result))
+        literal = node.children[0]
+        if not literal.datatype in [DataType.REAL, DataType.INTEGER]:
+            # The '-' operator can only be applied to REAL and INTEGER datatypes
```
Comment (arporter) [inline src/psyclone/psyir/backend/sir.py]: This line is too long.

File: src/psyclone/psyir/backend/sir.py [hunk 2]
```diff
@@ -403,10 +411,15 @@ def literal_node(self, node):
...
-        # #612).
-        return ("{0}make_literal_access_expr(\"{1}\", BuiltinType.Float)"
-                "".format(self._nindent, result))
+        try:
+            datatype = TYPE_MAP_TO_SIR[node.datatype]
+        except KeyError:
+            raise VisitorError(
+                "PSyIR type '{0}' has no representation in the SIR backend."
+                "".format(str(node.datatype)))
+
```
Comment (arporter) [inline src/psyclone/psyir/backend/sir.py]: Trailing whitespace.

File: src/psyclone/psyir/backend/sir.py (hunk 1 above)
Comment (rupertford) [reply src/psyclone/psyir/backend/sir.py]: Fixed, sorry.

File: src/psyclone/psyir/backend/sir.py (hunk 2 above)
Comment (rupertford) [reply src/psyclone/psyir/backend/sir.py]: And again. There was also a complaint about using not in the wrong place which I've also changed. pycodestyle and pylint are now happy.

**Comment by rupertford** [comment]
Ready for next review from @arporter

**Pull request merged**